File: src/core/splitter_dbs.py
"""
@package core
splitter_dbs module
"""

# -*- coding: iso-8859-15 -*-

# This code is distributed under the GNU General Public License (see
# THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
# Copyright (C) 2014, the P2PSP team.
# http://www.p2psp.org

# DBS: Data Broadcasting Set of rules

# {{{ Imports

#from __future__ import print_function
import threading
import sys
import socket
import struct
import time

<<<<<<< HEAD
from . import common
=======
from core.common import Common
>>>>>>> master
from core._print_ import _print_
from core.splitter_ims import Splitter_IMS
from core.color import Color

# }}}

# Some useful definitions.
ADDR = 0
PORT = 1

def _p_(*args, **kwargs):
    """Colorize the output."""
    if __debug__:
        sys.stdout.write(Common.DBS_COLOR)
        _print_("DBS:", *args)
        sys.stdout.write(Color.none)
    
class Splitter_DBS(Splitter_IMS):
    # {{{

    # {{{ Class "constants"

    MAX_CHUNK_LOSS = 32     # Chunk losses threshold to reject a peer from the team.
    MCAST_ADDR = "0.0.0.0"
    MONITOR_NUMBER = 1
    
    # }}}

    def __init__(self):
        # {{{

        Splitter_IMS.__init__(self) #???
        #sys.stdout.write(Color.yellow)
        #print("Using DBS")
        #sys.stdout.write(Color.none)

        #self.number_of_monitors = 0
        self.peer_number = 0

        # {{{ The list of peers in the team.
        # }}}
        self.peer_list = []

        # }}}

        # {{{ Destination peers of the chunk, indexed by a chunk
        # number. Used to find the peer to which a chunk has been
        # sent.
        # }}}
        self.destination_of_chunk = [('0.0.0.0',0)] * self.BUFFER_SIZE
        #for i in range(self.BUFFER_SIZE):
        #    self.destination_of_chunk.append(('0.0.0.0',0))

        self.losses = {}
        self.magic_flags = Common.DBS
        
        _p_("max_chunk_loss =", self.MAX_CHUNK_LOSS)
        _p_("mcast_addr =", self.MCAST_ADDR)
        _p_("Initialized")

        # }}}

        #def say_goodbye(self, node, sock):
        # {{{

        #sock.sendto(b'', node)

        # }}}

    def send_magic_flags(self, peer_serve_socket):

        message = struct.pack("B", self.magic_flags)
        peer_serve_socket.sendall(message)
        _print_("Magic flags =",bin(self.magic_flags))
            
    def send_the_list_size(self, peer_serve_socket):
        # {{{

        _p_("Sending the number of monitors", self.MONITOR_NUMBER)
        message = struct.pack("H", socket.htons(self.MONITOR_NUMBER))
        peer_serve_socket.sendall(message)
        _p_("Sending a list of peers of size", len(self.peer_list))
        message = struct.pack("H", socket.htons(len(self.peer_list)))
        peer_serve_socket.sendall(message)

        # }}}

    def send_the_list_of_peers(self, peer_serve_socket):
        # {{{

        #if __debug__:
        #    print("Sending a list of peers of size", len(self.peer_list))
        #message = struct.pack("H", socket.htons(len(self.peer_list)))
        #peer_serve_socket.sendall(message)
        self.send_the_list_size(peer_serve_socket)

        #peer_serve_socket.sendall(self.message)

        if __debug__:
            counter = 0
        for p in self.peer_list:
            message = struct.pack("4sH", socket.inet_aton(p[ADDR]), \
                                  socket.htons(p[PORT]))
            peer_serve_socket.sendall(message)
            if __debug__:
                _p_("[%5d]" % counter, p)
                counter += 1

        # }}}

    # def append_peer_borrame(self, peer):
    #     # {{{

    #     if peer not in self.peer_list:
    #         self.peer_list.append(peer)
    #     self.losses[peer] = 0

    #     print("DBS: ---------------------------------")

        # }}}

    def send_the_peer_endpoint(self, peer_serve_socket):
        # {{{

        peer_endpoint = peer_serve_socket.getpeername()
        message = struct.pack("4sH", socket.inet_aton(peer_endpoint[ADDR]), \
                              socket.htons(peer_endpoint[PORT]))
        peer_serve_socket.sendall(message)

        # }}}

    # Pensar en reutilizar Splitter_IMS.handle_peer_arrival()
    # concatenando las llamadas a las funciones.

    def send_configuration(self, sock):
        Splitter_IMS.send_configuration(self, sock)
        self.send_the_peer_endpoint(sock)
        self.send_magic_flags(sock)

    def insert_peer(self, peer):
        # {{{

        if peer not in self.peer_list: # Probar a quitar -----------------------------------------------------
            #self.peer_list.insert(self.peer_number, peer)
            self.peer_list.append(peer)
        self.losses[peer] = 0

        _p_("inserted peer", peer)

        # }}}

    def handle_a_peer_arrival(self, connection):
        # {{{

        # {{{ In the DBS, the splitter sends to the incomming peer the
        # list of peers. Notice that the transmission of the list of
        # peers (something that could need some time if the team is
        # big or if the peer is slow) is done in a separate thread. This
        # helps to avoid DoS (Denial of Service) attacks.
        # }}}

        serve_socket = connection[0]
        incomming_peer = connection[1]
        #sys.stdout.write(Color.green)
        #print(serve_socket.getsockname(), '\b: DBS: accepted connection from peer', \
        #      incomming_peer)
        _p_('accepted connection from peer', incomming_peer)
        #sys.stdout.write(Color.none)
        self.send_configuration(serve_socket)
        self.send_the_list_of_peers(serve_socket)
        serve_socket.close()
        self.insert_peer(incomming_peer)
        return incomming_peer

        # }}}

    def receive_message(self):
        # {{{

        try:
            return self.team_socket.recvfrom(struct.calcsize("H"))
        except:
            _p_("Unexpected error:", sys.exc_info()[0])
            pass

        # }}}

    def get_lost_chunk_number(self, message):
        # {{{

        return struct.unpack("!H",message)[0]

        # }}}

    def get_losser(self, lost_chunk_number):
        # {{{

        return self.destination_of_chunk[lost_chunk_number % self.BUFFER_SIZE]

        # }}}

    def remove_peer(self, peer):
        # {{{

        try:
            self.peer_list.remove(peer)
        except ValueError:
            pass
        else:
            self.peer_number -= 1

        try:
            del self.losses[peer]
        except KeyError:
            pass

        # }}}

    def increment_unsupportivity_of_peer(self, peer):
        # {{{

        try:
            self.losses[peer] += 1
        except KeyError:
            _p_("the unsupportive peer", peer, "does not exist!")
        else:
            #sys.stdout.write(Color.blue)
            _p_(peer, "has loss", self.losses[peer], "chunks")
            #sys.stdout.write(Color.none)
            if self.losses[peer] > self.MAX_CHUNK_LOSS:
                if peer not in self.peer_list[:self.MONITOR_NUMBER]:
                    #sys.stdout.write(Color.red)
                    _p_(peer, 'removed')
                    self.remove_peer(peer)
                    #sys.stdout.write(Color.none)
        finally:
           pass

        # }}}

    def process_lost_chunk(self, lost_chunk_number, sender):
        # {{{

        destination = self.get_losser(lost_chunk_number)

        if __debug__:

            #sys.stdout.write(Color.cyan)
            _p_(sender, "complains about lost chunk", lost_chunk_number, "sent to", destination)
            #sys.stdout.write(Color.none)

            if destination in self.peer_list[:self.MONITOR_NUMBER]:
                _p_("lost chunk index =", lost_chunk_number)

        self.increment_unsupportivity_of_peer(destination)

        # }}}

    def process_goodbye(self, peer):
        # {{{

        #sys.stdout.write(Color.green)
        _p_('Received "goodbye" from', peer)
        #sys.stdout.write(Color.none)
        sys.stdout.flush()

        #if peer not in self.peer_list[:self.MONITOR_NUMBER]:
        self.remove_peer(peer)

        # }}}

    def moderate_the_team(self):
        # {{{

        while self.alive:
            # {{{

            message, sender = self.receive_message()

            if len(message) == 2:

                # {{{ The peer complains about a lost chunk.

                # In this situation, the splitter counts the number of
                # complains. If this number exceeds a threshold, the
                # unsupportive peer is expelled from the
                # team.

                lost_chunk_number = self.get_lost_chunk_number(message)
                self.process_lost_chunk(lost_chunk_number, sender)

                # }}}

            else:

                # {{{ The peer wants to leave the team.

                # A !2-length payload means that the peer wants to go
                # away.

                self.process_goodbye(sender)

                # }}}

            # }}}

        # }}}


    def setup_team_socket(self):
        # {{{

        # "team_socket" is used to talk to the peers of the team.
        self.team_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

        # This does not work in Windows systems !!
        self.team_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        #self.team_socket.settimeout(1000.0)
        self.team_socket.bind(('', self.PORT))

        # }}}

    def reset_counters(self):
        # {{{

        for i in self.losses:
            self.losses[i] /= 2

        # }}}

    def reset_counters_thread(self):
        # {{{

        while self.alive:
            self.reset_counters()
            time.sleep(Common.COUNTERS_TIMING)

        # }}}

    def compute_next_peer_number(self, peer):
        # {{{ 

        self.peer_number = (self.peer_number + 1) % len(self.peer_list)

        # }}}

    def run(self):
        # {{{

        self.receive_the_header()

        # {{{ A DBS splitter runs 4 threads. The main one and the
        # "handle_arrivals" thread are equivalent to the daemons used
        # by the IMS splitter. "moderate_the_team" and
        # "reset_counters_thread" are new.
        # }}}

        _p_(self.peer_connection_socket.getsockname(), "\b: waiting for the monitor peers ...")
        def _():
            connection  = self.peer_connection_socket.accept()
            incomming_peer = self.handle_a_peer_arrival(connection)
            #self.insert_peer(incomming_peer) # Notice that now, the
                                             # monitor peer is the
                                             # only one in the list of
                                             # peers. It is no
                                             # neccesary a delay.
        _()

        threading.Thread(target=self.handle_arrivals).start()
        threading.Thread(target=self.moderate_the_team).start()
        threading.Thread(target=self.reset_counters_thread).start()

        message_format = self.chunk_number_format \
                        + str(self.CHUNK_SIZE) + "s"

        #header_load_counter = 0
        while self.alive:

            chunk = self.receive_chunk()
            try:
                peer = self.peer_list[self.peer_number]
                message = struct.pack(message_format, socket.htons(self.chunk_number), chunk)
                self.send_chunk(message, peer)

                self.destination_of_chunk[self.chunk_number % self.BUFFER_SIZE] = peer
                self.chunk_number = (self.chunk_number + 1) % Common.MAX_CHUNK_NUMBER
                self.compute_next_peer_number(peer)
            except IndexError:
                _p_("The monitor peer has died!")


        # }}}

    # }}}

File: src/core/splitter_dbs.cc
//
//  splitter_dbs.cc -- Data Broadcasting Set of rules
//
//  This code is distributed under the GNU General Public License (see
//  THE_GENERAL_GNU_PUBLIC_LICENSE.txt for extending this information).
//  Copyright (C) 2016, the P2PSP team.
//  http://www.p2psp.org
//

#include "splitter_dbs.h"
#include "../util/trace.h"

namespace p2psp {
  using namespace std;
  using namespace boost;

  const int Splitter_DBS::kMaxChunkLoss = 32;  // Chunk losses threshold to reject a peer from the team
  const int Splitter_DBS::kNumberOfMonitors = 1;

  Splitter_DBS::Splitter_DBS() : Splitter_core(), losses_(0, &Splitter_DBS::GetHash) {
    // TODO: Check if there is a better way to replace kMcastAddr with 0.0.0.0
    max_number_of_chunk_loss_ = kMaxChunkLoss;
    number_of_monitors_ = kNumberOfMonitors;

    peer_number_ = 0;
    destination_of_chunk_.reserve(buffer_size_);
#if defined __DEBUG__ || defined __PARAMS__
    TRACE("max_number_of_chunk_loss = "
	  << max_number_of_chunk_loss_);
#endif
#if defined __DEBUG__ || defined __SORS__
    TRACE("Splitter_DBS constructor");
#endif
  }

  Splitter_DBS::~Splitter_DBS() {}

  void Splitter_DBS::SendTheNumberOfPeers(const std::shared_ptr<boost::asio::ip::tcp::socket> &peer_serve_socket) {
    char message[2];

#if defined __DEBUG__ || defined __CHURN__
    TRACE("Sending the number of monitors "
	  << number_of_monitors_);
#endif
    (*(uint16_t *)&message) = htons(number_of_monitors_);
    peer_serve_socket->send(asio::buffer(message));
    
#if defined __DEBUG__ || defined __CHURN__
    TRACE("Sending a list of peers of size "
	  << to_string(peer_list_.size()));
#endif
    (*(uint16_t *)&message) = htons(peer_list_.size());
    peer_serve_socket->send(asio::buffer(message));
  }

  void Splitter_DBS::SendTheListOfPeers(const std::shared_ptr<boost::asio::ip::tcp::socket> &peer_serve_socket) {
    SendTheNumberOfPeers(peer_serve_socket);

    int counter = 0;
    char message[6];
    in_addr addr;

#if defined __DEBUG__ || defined __CHURN__
    TRACE("Peer list length = "
	  << peer_list_.size());
#endif
    
    for (std::vector<asio::ip::udp::endpoint>::iterator it = peer_list_.begin();
         it != peer_list_.end(); ++it) {
      inet_aton(it->address().to_string().c_str(), &addr);
      (*(in_addr *)&message) = addr;
      (*(uint16_t *)(message + 4)) = htons(it->port());
      peer_serve_socket->send(asio::buffer(message));

#if defined __DEBUG__ || defined __CHURN__
      TRACE(to_string(counter)
	    << ", "
	    << *it);
#endif      
      counter++;
    }
  }

  /*void Splitter_DBS::SendThePeerEndpoint(const std::shared_ptr<boost::asio::ip::tcp::socket> &peer_serve_socket) {
    asio::ip::tcp::endpoint peer_endpoint = peer_serve_socket->remote_endpoint();

    char message[6];
    in_addr addr;
    inet_aton(peer_endpoint.address().to_string().c_str(), &addr);
    (*(in_addr *)&message) = addr;
    (*(uint16_t *)(message + 4)) = htons(peer_endpoint.port());
    peer_serve_socket->send(asio::buffer(message));
    }*/

  void Splitter_DBS::InsertPeer(const boost::asio::ip::udp::endpoint &peer) {
    if (find(peer_list_.begin(), peer_list_.end(), peer) != peer_list_.end()) {
      peer_list_.erase(find(peer_list_.begin(), peer_list_.end(), peer));
    }
    peer_list_.push_back(peer);
    losses_[peer] = 0;
#if defined __DEBUG__ || defined __CHURN__
    TRACE("Inserted peer "
	  << peer);
#endif
  }

  void Splitter_DBS::SendConfiguration(const std::shared_ptr<boost::asio::ip::tcp::socket> &sock) {
#if defined __DEBUG__ || defined __CHURN__
    TRACE("Sending Configuration");
#endif
    Splitter_core::SendConfiguration(sock);
    SendTheListOfPeers(sock);

    //SendThePeerEndpoint(sock);
  }

  void Splitter_DBS::HandleAPeerArrival(std::shared_ptr<boost::asio::ip::tcp::socket> serve_socket) {
    /* In the DBS, the splitter sends to the incomming peer the
       list of peers. Notice that the transmission of the list of
       peers (something that could need some time if the team is
       big or if the peer is slow) is done in a separate thread. This
       helps to avoid DoS (Denial of Service) attacks.
    */

    asio::ip::tcp::endpoint incoming_peer = serve_socket->remote_endpoint();

#if defined __DEBUG__ || defined __CHURN__
    TRACE("Accepted connection from peer "
	  << incoming_peer);
#endif
    
    SendConfiguration(serve_socket);
    //SendTheListOfPeers(serve_socket);
    ReceiveReadyForReceivingChunks(serve_socket);
    serve_socket->close();
    boost::asio::ip::udp::endpoint incoming_peer_udp(incoming_peer.address(), incoming_peer.port());
    InsertPeer(incoming_peer_udp);

    // TODO: In original code, incoming_peer is returned, but is not used
  }

  size_t Splitter_DBS::ReceiveMessage(std::vector<char> &message, boost::asio::ip::udp::endpoint &endpoint) {
    system::error_code ec;

    size_t bytes_transferred =
      team_socket_.receive_from(asio::buffer(message), endpoint, 0, ec);

    if (ec) {
      ERROR("Unexepected error: "
	    << ec.message());
    }

    return bytes_transferred;
  }

  void Splitter_DBS::IncrementPeerUnsupportivity(const boost::asio::ip::udp::endpoint &peer) {
    bool peerExists = true;

    try {
      losses_[peer] += 1;
    } catch (std::exception e) {
      TRACE("The unsupportive peer "
	    << peer
	    << " does not exist!");
      peerExists = false;
    }

    if (peerExists) {

#if defined __DEBUG__ || defined __LOST_CHUNKS__
      TRACE(""
	    << peer
	    << " has lost "
	    << to_string(losses_[peer])
	    << " chunks");
#endif
      
      if (losses_[peer] > max_number_of_chunk_loss_) {

#if defined __DEBUG__ || defined __LOST_CHUNKS__
	TRACE(""
	      << peer
	      << " removed by unsupportive");
#endif
	
        RemovePeer(peer);
      }
    }
  }

  void Splitter_DBS::ProcessLostChunk(int lost_chunk_number, const boost::asio::ip::udp::endpoint &sender) {
    asio::ip::udp::endpoint destination = GetLosser(lost_chunk_number);

#if defined __DEBUG__ || defined __LOST_CHUNKS__
    TRACE(""
	  << sender
	  << " complains about lost chunk "
          << to_string(lost_chunk_number)
	  << " sent to "
	  << destination);

    if (find(peer_list_.begin() + number_of_monitors_, peer_list_.end(), destination) != peer_list_.end()) {
      TRACE("Lost chunk index = "
	    << lost_chunk_number);
    }
#endif

    IncrementPeerUnsupportivity(destination);
  }

  uint16_t Splitter_DBS::GetLostChunkNumber(const std::vector<char> &message) {
    // TODO: Check if this is totally correct
    return ntohs(*(uint16_t *)message.data());
  }

  asio::ip::udp::endpoint Splitter_DBS::GetLosser(int lost_chunk_number) {
    return destination_of_chunk_[lost_chunk_number % buffer_size_];
  }

  void Splitter_DBS::RemovePeer(const asio::ip::udp::endpoint &peer) {
    // If peer_list_ contains the peer, remove it
    if (find(peer_list_.begin(), peer_list_.end(), peer) != peer_list_.end()) {
      peer_list_.erase(remove(peer_list_.begin(), peer_list_.end(), peer), peer_list_.end());

      // In order to avoid negative peer_number_ value while peer_list_ still
      // contains any peer (in Python this is not necessary because negative
      // indexes can be used)
      if (peer_list_.size() > 0) {
        peer_number_ = (peer_number_ - 1) % peer_list_.size();
      } else {
        peer_number_--;
      }
    }

    losses_.erase(peer);
  }

  void Splitter_DBS::ProcessGoodbye(const boost::asio::ip::udp::endpoint &peer) {
#if defined __DEBUG__ || defined __CHURN__
    TRACE("Received 'goodbye' from "
	  << peer);
#endif
    
    // TODO: stdout flush?

    if (find(outgoing_peer_list_.begin(), outgoing_peer_list_.end(),peer) == outgoing_peer_list_.end()){
      if (find(peer_list_.begin(), peer_list_.end(), peer) != peer_list_.end()){
	outgoing_peer_list_.push_back(peer);
#if defined __DEBUG__ || defined __CHURN__
	TRACE("Marked for deletion: " << peer);
#endif
      }
    }


    //SayGoodbye(peer);
    //RemovePeer(peer);

  }

  void Splitter_DBS::SayGoodbye(const boost::asio::ip::udp::endpoint &peer) {
    std::string goodbye("G");
    team_socket_.send_to(boost::asio::buffer(goodbye), peer);

#if defined __DEBUG__ || defined __CHURN__
    TRACE("[Goodbye] sent to "
	  << "("
	  << peer.address().to_string()
	  << ","
	  << std::to_string(peer.port())
	  << ")");
#endif
    
  }

  void Splitter_DBS::ModerateTheTeam() {
    std::vector<char> message(2);
    asio::ip::udp::endpoint sender;

    while (alive_) {
      size_t bytes_transferred = ReceiveMessage(message, sender);

      if (bytes_transferred == 2) {
        /*
          The peer complains about a lost chunk.

          In this situation, the splitter counts the number of
          complains. If this number exceeds a threshold, the
          unsupportive peer is expelled from the
          team.
        */

        uint16_t lost_chunk_number = GetLostChunkNumber(message);
        ProcessLostChunk(lost_chunk_number, sender);

      } else {
        /*
          The peer wants to leave the team.

          A !2-length payload means that the peer wants to go
          away.
        */

        ProcessGoodbye(sender);
      }
    }
  }

  void Splitter_DBS::SetupTeamSocket() {
    system::error_code ec;
    asio::ip::udp::endpoint endpoint(asio::ip::udp::v4(), splitter_port_);

    team_socket_.open(asio::ip::udp::v4());

    asio::socket_base::reuse_address reuseAddress(true);
    team_socket_.set_option(reuseAddress, ec);
    team_socket_.bind(endpoint);

    if (ec) {
      ERROR(ec.message());
    }
  }

  void Splitter_DBS::ResetCounters() {
    unordered::unordered_map<asio::ip::udp::endpoint, int>::iterator it;
    for (it = losses_.begin(); it != losses_.end(); ++it) {
      losses_[it->first] = it->second / 2;
    }
  }

  void Splitter_DBS::ResetCountersThread() {
    while (alive_) {
      ResetCounters();
      this_thread::sleep(posix_time::seconds(Common::kCountersTiming));
    }
  }

  void Splitter_DBS::ComputeNextPeerNumber(asio::ip::udp::endpoint &peer) {
    if (peer_list_.size() > 0)
      peer_number_ = (peer_number_ + 1) % peer_list_.size();
  }

  void Splitter_DBS::Run() {
    ConfigureSockets();
    RequestTheVideoFromTheSource();

    /* A DBS splitter runs 4 threads. The main one and the
       "handle_arrivals" thread are equivalent to the daemons used
       by the IMS splitter. "moderate_the_team" and
       "reset_counters_thread" are new.
    */

#if defined __DEBUG__ || defined __CHURN__
    TRACE("waiting for the monitor peers ...");
#endif
    
    std::shared_ptr<asio::ip::tcp::socket> connection = make_shared<asio::ip::tcp::socket>(boost::ref(io_service_));
    acceptor_.accept(*connection);
    HandleAPeerArrival(connection);

    // Threads
    thread t1(bind(&Splitter_core::HandleArrivals, this));
    thread t2(bind(&Splitter_DBS::ModerateTheTeam, this));
    thread t3(bind(&Splitter_DBS::ResetCountersThread, this));

    vector<char> message(sizeof(uint16_t) + chunk_size_);
    asio::ip::udp::endpoint peer;

    while (alive_) {
      asio::streambuf chunk;
      size_t bytes_transferred = ReceiveChunk(chunk);
      try {
        peer = peer_list_.at(peer_number_);

        (*(uint16_t *)message.data()) = htons(chunk_number_);

        copy(asio::buffer_cast<const char *>(chunk.data()),
             asio::buffer_cast<const char *>(chunk.data()) + chunk.size(), message.data() + sizeof(uint16_t));

        SendChunk(message, peer);

        destination_of_chunk_[chunk_number_ % buffer_size_] = peer;
        chunk_number_ = (chunk_number_ + 1) % Common::kMaxChunkNumber;
        ComputeNextPeerNumber(peer);
      } catch (const std::out_of_range &oor) {
        ERROR("The monitor peer has died!");
        exit(-1);
      }

      if (peer_number_ == 0){
	for (unsigned int i=0; i<outgoing_peer_list_.size(); i++){
	  SayGoodbye(outgoing_peer_list_[i]);
	  RemovePeer(outgoing_peer_list_[i]);
	}
	outgoing_peer_list_.clear();
      }
      
      chunk.consume(bytes_transferred);
    }
  }

  std::vector<boost::asio::ip::udp::endpoint> Splitter_DBS::GetPeerList() {
    return peer_list_;
  }

  int Splitter_DBS::GetLoss(const boost::asio::ip::udp::endpoint &peer) {
    return losses_[peer];
  }

  void Splitter_DBS::SetMaxNumberOfChunkLoss(int max_number_of_chunk_loss) {
    max_number_of_chunk_loss_ = max_number_of_chunk_loss;
  }

  int Splitter_DBS::GetMaxNumberOfChunkLoss() {
    return max_number_of_chunk_loss_;
  }

  int Splitter_DBS::GetDefaultMaxNumberOfChunkLoss() {
    return kMaxChunkLoss;
  }

  void Splitter_DBS::SetNumberOfMonitors(int number_of_monitors) {
    number_of_monitors_ = number_of_monitors;
  }

  int Splitter_DBS::GetNumberOfMonitors() {
    return number_of_monitors_;
  }

  int Splitter_DBS::GetDefaultNumberOfMonitors() {
    return kNumberOfMonitors;
  }

  void Splitter_DBS::Start() {
    thread_.reset(new boost::thread(boost::bind(&Splitter_DBS::Run, this)));
  }

}
